datalad/core/local/run.py

```diff
@@ -394,17 +395,28 @@ def _unlock_or_remove(dset_path, paths):
             lgr.debug("Filtered out non-existing path: %s", path)
 
     if existing:
-        remove = Remove()
+        notpresent_results = []
+        # Note: If Unlock() is given a directory (including a subdataset) as a
+        # path, files without content present won't be reported, so those cases
+        # aren't being covered by the "remove if not present" logic below.
         for res in Unlock()(dataset=dset_path, path=existing,
                             on_failure="ignore"):
-            if res["status"] == "impossible":
-                if "cannot unlock" in res["message"]:
-                    for rem_res in remove(dataset=dset_path,
-                                          path=res["path"],
-                                          check=False, save=False):
-                        yield rem_res
-                    continue
+            if res["status"] == "impossible" and res["type"] == "file" \
+               and "cannot unlock" in res["message"]:
+                notpresent_results.append(res)
+                continue
             yield res
+        # Avoid `datalad remove` because it calls git-rm underneath, which will
+        # remove leading directories if no other files remain. See gh-5486.
+        for res in notpresent_results:
+            try:
+                os.unlink(res["path"])
+            except OSError as exc:
+                yield dict(res, action="run-remove", status="error",
+                           message=("Removing file failed: %s", exc_str(exc)))
+            else:
+                yield dict(res, action="run-remove", status="ok",
```

Comment:
just an idea/observation, feel free to ignore:
our "action" values are indeed nowhere "standardized" although could be relied upon in the logic "upstairs".  Typically they correspond to the actual "full" action (like `run` here)

<details>
<summary>here is what I see values ATM</summary> 

```shell
$> git grep "\<action=[\"']" | grep -v 'form action' | sed -e "s,.*\<action=['\"]\([^\"']*\)['\"].*,\1,g" | sort -n | uniq -c | sort -n
      1 add-sibling
      1 addurl
      1 bar
      1 clean
      1 create_sibling
      1 delete
      1 discover_procedure
      1 dummy
      1 enable
      1 enable-sibling
      1 export_archive
      1 export-archive-ora
      1 export_to_figshare
      1 fetch
      1 funky
      1 no_annex
      1 push
      1 query-sibling
      1 remove-sibling
      1 run_procedure
      1 store
      1 subdataset
      1 test
      1 version
      2 add_readme
      2 add_submodule
      2 copy_file
      2 foo
      2 procedure_help
      2 uninstall
      2 wtf
      3 create
      3 drop
      3 search
      4 query
      4 remove
      4 run
      5 aggregate_metadata
      5 configure-sibling
      6 create_sibling_gitlab
      7 status
      7 unlock
      8 create-sibling-ria
      8 merge
      9 addurls
      9 metadata
     11 store_false
     12 install
     13 add
     13 save
     18 get
     21 append
     22 publish
     29 update
     32 copy
     34 diff
     69 store_true
```
</details> 
so typically correspond to the actual user facing "action" and typically in Python syntax (with `_` as separator), although have `-` in some cases (`export-archive-ora`) and also freeform new minted action names like `query-sibling` similar to here `run-remove` which do not correspond to any command.  So adding a new one here does not really break any consistency any longer, but may be it could be a good time to introduce some notion of "subaction".  I.e. here it could be `run.remove` or `run:remove` to signal that it is `remove` step of `run` action, so that later code upstairs could start using such syntax to separate out action (`run`) from subaction (`remove`) in a consistent manner?

Reply:
> Typically they correspond to the actual "full" action (like `run` here)

Right.  I went with "run" first, but switched to something more specific because I was worried it'd cause confusion.

> So adding a new one here does not really break any consistency any longer, but may be it could be a good time to introduce some notion of "subaction". I.e. here it could be `run.remove` or `run:remove` to signal that it is `remove` step of `run` action, so that later code upstairs could start using such syntax to separate out action (`run`) from subaction (`remove`) in a consistent manner?

I think both your suggestions ("run.remove" and "run:remove") are better than what I chose.  I'll switch it over to "run.remove".